### ISSUE TITLE
HttpContextBase shouldn't be injected or resolved directly

### ISSUE THREAD

**Issue by orchardbot** [OrchardCMS/Orchard#4607]
@Piedone created:
https://orchard.codeplex.com/workitem/20778

We have a single implementation to provide if we want to change HttpContext behaviour, and that's by implementing IHttpContextAccessor.

Now in MvcModule there is a direct registration for (among others) HttpContextBase. This can give one an impression that it's fine to resolve/inject that directly anywhere where one needs to access the HttpContext. However this is not true and this should rather happen only for background tasks.

We should do something do provide an IHttpContextAccessor providing the mockup HttpContext in background tasks (we could have a separate IHttpContextAccessor implementation and register it for the life time scope of background tasks only). Otherwise HttpContextBase should never be injected because it just makes things more complicated when implementations should be changed (e.g. ShapeTemplateBindingStrategy ln 179 and ShapeDisplay ln 19 uses HttpContextBase directly).

**Comment by sfmskywalker** [comment]
(Copied part of my comment on #5306 here since it really was a comment about this issue) 

Why do we think it's bad practice to inject ```HttpContextBase``` instead of an interface? It seems to me that the very reason for the existence of ```HttpContextBase``` is to be able to substitute different implementations of it, like interfaces. The only difference is that it is an abstract class instead of an interface, but in principle it's the same pattern.

Thoughts?

**Comment by Piedone** [comment]
Swapping out implementation of IHttpContextAccessor and change it to use an arbitrary logic is easier (even though it's an app-wide singleton) than to do the same with HttpContextBase: IHttpContextAccessor.Current() can contain arbitrary logic that will run when the HttpContext is needed (like creating a fake context when HttpContext.Current is null), and while you can implement similar logic from Autofac event handlers for a HttpContextBase resolution it's more complicated, less intuitive and special in a way that it's not what we do in case of other dependencies.
Furthermore if we use both IHttpContextAccessor and HttpContextBase then there are two services to override.
Now you have to know whether your code will ever run in a bg task (what you can often not tell in advance) and then use HttpContextBase and if not then IHttpContextAccessor. Instead you'd just need to use IHttpContextAccessor.

**Comment by sfmskywalker** [comment]
Alternatively, we do change the implementation of ```IHttpContextAccessor``` such that it always returns either an ```HttpContextWrapper``` of an ```HttpContext.Current```, or if ```HttpContext.Current``` is null, an instance of ```HttpContextPlaceholder```. Next we'd update the 2 places in Orchard where we need to know if we're in a background task or not (```WorkContextAccessor``` and ```MvcModule```).

**Comment by sfmskywalker** [comment]
Please review my changes here: https://github.com/OrchardCMS/Orchard/commit/ef2abe14cbb35b3f42c8a1b6f9c606b1c299071d

Note that I did not change any behavior of the ```IHttpContextAccessor``` - all I did in principle is initializing ```HttpContext.Current``` with an actual instance during background sweeps so that Html Helpers won't fail when Razor views are being rendered in the background.

**Comment by sfmskywalker** [comment]
@Zoltan Good points. Then let's change the implementation of ```IHttpContextAccessor``` to always return an actual ```HttpContextBase``` instance instead of null, and update code elsewhere that is currently checking for ```IHttpContextAccessor.Current()``` being null appropriately.

**Comment by zoltan** [comment]
again wrong Zoltan...

**Comment by sfmskywalker** [comment]
@zoltan Sorry.

**Comment by sfmskywalker** [comment]
Looks like we won't be able to use the implementation of ```IHttpContextAccessor``` because it is registered with the container using the "SingleInstance" scope, where the ```HttpContextBase``` is registered using the "InstancePerRequest" scope, so we better keep those scopes as-is, or register ```IHttpContextAccessor``` to use the transient scope as well. Otherwise we'd have to come up with another interface that provides access to ```HttpContextBase```. Something like ```IHttpContextWrapper``` for example. The downside of that is that it will be confusing to understand when to use what.

**Comment by Piedone** [comment]
Given that the HttpContextBase and RequestContext registered in MvcModule are only interesting for background tasks (as otherwise the same information is in HttpContet.Current) we could provide an alternate implementation for IHttpContextAccessor just when running background tasks. The same happens in Orchard Application Host, see: http://orchardapphost.codeplex.com/SourceControl/latest#OrchardAppHost.cs from "var hca = new AppHostHttpContextAccessor();".

This way regardless of IHttpContextAccessor being an app-wide singleton it can be overridden to have a separate implementation for background tasks (and that can even be an InstancePerDependency as the ones in MvcModule, though I don't know why that would be needed; maybe one instance per scope, but one instance per dependency?) that would contain the same logic as MvcModule in its Current() method.

**Comment by sfmskywalker** [comment]
I like the notion of a background-specific implementation when executing in a background task, but I fail to see how to dynamically register such a component when executing from a background task. One way I see is registering a factory method with the container as is done in MvcModule for HttpContextBase. Is that what you mean as well, or something else?

**Comment by sfmskywalker** [comment]
After doing some experiments and failing to make things work with an additional IHttpContextAccessor implementation for background tasks due to lifetimescope mismatch issues, the following changes will work flawlessly:

1. Change the ```HttpContextAccessor``` implementation to return the HttpContextBase by resolving it like so:

```C#
public class HttpContextAccessor : IHttpContextAccessor {
        private readonly Lazy<IComponentContext>_context;

        public HttpContextAccessor(Lazy<IComponentContext> context) {
            _context = context;
        }

        public HttpContextBase Current() {
            HttpContextBase httpContextBase;
            _context.Value.TryResolve(out httpContextBase);
            return httpContextBase;
        }
    }
```

2. All code using ```IHttpContextAccessor``` that retrieves the current HttpContext using the ```.Current()``` method and checking for it to be ```null``` will be updated to check for the HttpContext being a "background http context" instead using an extension method on ```HttpContextBase``` called ```IsBackgroundContext```:

```C#
public static bool IsBackgroundContext(this HttpContextBase httpContextBase) {
   return httpContextBase == null || httpContextBase is MvcModule.HttpContextPlaceholder;
}
```

This implementation has low risk of breaking things, is simple, and enables us to inject IHttpContextAccessor and have it return a valid HttpContextBase instance (without having to inject HttpContextBase directly), even in background tasks.

The ```IsBackgroundContext``` check is only necessary for certain parts in Orchard where it needs to be aware of its context (being background or not) and is a better implementation than checking for null anyway.

I'll push the changes to the issue branch for your review and comments.

**Comment by Piedone** [comment]
I like the solution, especially the explicit IsBackgroundContext(), it came to my mind too.

So you tried the BeginLifetimeScope() trick as in the linked class of Orchard App Host too and it didn't work?

**Comment by sfmskywalker** [comment]
Thanks! Good to hear.

I didn't try the ```BeginLifetimeScope``` trick because, well, it's a trick. ;)
My current suggestion I think is both simple and elegant without changing too much.

However, if you feel that the ```BeginLifeTimeScope()``` thing with an additional ```IHttpContextAccessor``` implementation is more suitable, please do share a working version of that - it would be interesting to see at the very least.

**Comment by Piedone** [comment]
It's not that of a trick, I mean it's not a hack but a perfectly supported legitimate scenario. But I won't spend time on implementing it since I know how it looks :-) and I don't see any advantages apart from not touching anything else than SweepGenerator.

**Comment by sfmskywalker** [comment]
Fair enough. :)

**Comment by sfmskywalker** [comment]
Fixed in #34008c469eebd65829885bde1973e02cbbe1cbe5
